### ISSUE TITLE
Question/request about representative document truncation 

### ISSUE THREAD

**Issue by zilch42** [MaartenGr/BERTopic#1527]
Hi Maarten,

I'm curious about the truncation at 255 characters of representative documents in both the `OpenAI` and `TextGeneration` representation models (at least). 

https://github.com/MaartenGr/BERTopic/blob/58d90bc65e95a11718e63a0834809d156dcf431d/bertopic/representation/_textgeneration.py#L143C1-L147C67

I imagine it's helpful if you are paying for an API, but if using a local text generation model e.g. as in your Llama2 google colab example, is there any other reason that the representative documents have to be cut down so hard? 

Could we have some control over the truncation e.g. add a `truncate: int = 255` parameter to the representation models and then:

```
for doc in docs:
    doc = doc[:truncate] if truncate else doc
    to_replace += f"- {doc}\n"
```

**Comment by MaartenGr** [comment]
Yeah, I introduced that when LLMs had relatively small context sizes and as the trained documents are often relatively big (compared to context sizes) this was an easy limit to just use. Llama 2 models also have context sizes that limit the amount of text that you can give it. So you have to be really careful to not introduce too many tokens for just the documents. 

Having said that, opening up a parameter would be no problem here. I would actually prefer to have a truncate parameter that relates to the underlying tokenizer but unfortunately that is not easily possible without introducing many dependencies. Do you think it would suffice to use a character-based truncate parameter?

**Comment by zilch42** [comment]
Yeah I he was thinking about that too. Is it generally true that when using `TextGeneration`, the user should have the tokenizer for the text generation model they want to use? In that case, the dependency is on the user not on BERTopic. 

Does `tiktoken` have a lot of dependencies? That covers `OpenAI` but I'm not sure what cohere uses. 

If not characters, would words be better because it's a bit easier for people to estimate tokens from words than it is from characters? E.g. for OpenAI 100 tokens ~ 75 words. It might also be useful too add to the documentation how many tokens the standard prompts e.g. for `OpenAI` amount to. Then you could have a default of maybe 50 words rather than 255 characters for shorter context length generation models.

**Comment by MaartenGr** [comment]
> Yeah I was thinking about that too. Is it generally true that when using TextGeneration, the user should have the tokenizer for the text generation model they want to use? In that case, the dependency is on the user not on BERTopic and TextGeneration just needs to ask for it.
> 
> Does tiktoken have a lot of dependencies? That covers OpenAI but I'm not sure what cohere uses.

I think it will be a bit more difficult than only the dependencies. Each LLM uses a different tokenizer and for each LLM, that specific tokenizer should be used. They should generally work the same in their respective classes (`TextGeneration`, `OpenAI`, `Cohere`) but that will not always be the case especially if one uses a more broader implementation, like LangChain. So I am not that sure whether all tokenizers can easily be accessed. 

> If not characters, would words be better because it's a bit easier for people to estimate tokens from words than it is from characters? E.g. for OpenAI 100 tokens ~ 75 words. It might also be useful too add to the documentation how many tokens the standard prompts e.g. for OpenAI amount to. Then you could have a default of maybe 50 words rather than 255 characters for shorter context length generation models.

Splitting into words is only possible if they are separated with whitespaces which is not the case for many non-western languages.

**Comment by zilch42** [comment]
Ah I see. LangChain and non english languages aren't things I've dived into yet. That does pose a challenge. Hmmm my only suggestion would be to have more options and handle different cases e.g. 

```
truncate_at: int = 255,
split_on: str = ['char', 'whitespace', 'token'],
tokenizer = None,

```

But that may be a lot of work and cases to handle. If characters is the only option for now, that would still be a good improvement!

**Comment by MaartenGr** [comment]
I think that would create a bit more bloat with respect to the parameter space. I want to keep it as tight as possible to optimize an intuitive interface/experience. Perhaps only keep `char` and `token` and skip the `whitespace`? We could also make it so that users choose what to split on. So instead of `'whitespace'` it would be `' '`.

**Comment by zilch42** [comment]
Yep that would make sense, if it took the same input as `.split()`

**Comment by MaartenGr** [comment]
@zilch42 I just created a pull request that implements these features with two variables, `doc_length` and `tokenizer` in #1539. If you have the time to try it out, let me know what you think!
